File: persistance/src/main/java/ro/mpp2024/AbstractRepository.java
package ro.mpp2024;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.utils.JdbcUtils;

import java.util.Optional;
import java.util.Properties;

public abstract class AbstractRepository<ID extends Comparable<ID>, E extends Entity<ID>> implements  IRepository<ID, E> {
    protected JdbcUtils jdbc;
    protected static final Logger logger= LogManager.getLogger();

    public AbstractRepository(Properties props) {
        jdbc=new JdbcUtils(props);
    }

    @Override
    public abstract Optional<E> findById(ID id);

    @Override
    public abstract Iterable<E> findAll();

    @Override
    public abstract Optional<E> delete(ID id);

    @Override
    public abstract Optional<E> update(E entity);
}

File: csharpServer/persistance/AbstractRepository.cs
using System;
using System.Collections.Generic;
using Avalonia.Data;
using Microsoft.Extensions.Logging;
using model;

namespace persistance;

public abstract class AbstractRepository<ID, E> : IRepository<ID, E> where ID : IComparable<ID> where E : Entity<ID>
{
    protected JdbcUtils jdbc;
    protected static readonly ILogger logger = LoggerFactory.Create(builder => builder.AddConsole()).CreateLogger<AbstractRepository<ID, E>>();

    protected AbstractRepository()
    {
        // Initialize JdbcUtils with configuration from appsettings.json
        jdbc = new JdbcUtils();
    }

    public abstract Optional<E> FindById(ID id);

    public abstract IEnumerable<E> FindAll();
    
    public abstract Optional<E> Save(E entity);

    public abstract Optional<E> Delete(ID id);

    public abstract Optional<E> Update(E entity);
}